### ISSUE TITLE
Enable `ocamlnat` on the testsuite

### ISSUE THREAD

**Issue by Armael** [ocaml/ocaml#8989]
The current testsuite does not test ocamlnat (the native toplevel) at all (leading to issues such as #8988 to go undetected).

Support for ocamlnat in "toplevel" tests is implemented, but not enabled. To enable it, one should uncomment the ocamlnat-related lines in `ocamltest/ocaml_test.ml` (in the definition of `toplevel`).

Then, the second part of the issue is that it is likely that several of the toplevel tests in the testsuite do not currently pass with ocamlnat (due to bugs in ocamlnat or discrepancies with the standard toplevel).

We should probably investigate these cases, and either fix them or have a way in ocamltest to specify that the output is supposed to be different for the "standard toplevel" and "native toplevel".

cc @shindere

**Comment by shindere** [comment]
Thanks a lot for having taken the time to submit this issue.

A few hints:

To find the toplevel tests in the testsuite, go to the `testsuite/tests`
directory and from there do:

```bash
git grep "* toplevel"
```

When a test `foo.ml` is expected to produce the same output when run
under `ocaml` and `ocamlnat`, then the reference file is supposed to be
called `foo.compilers.reference`. When `ocaml` and `ocamlnat` are
expected to produce different results, then the corresponding files
should be named `foo.ocaml.reference` and `foo.ocamlnat.reference`.

**Comment by gasche** [comment]
When we resurrected `ocamlnat`, the general agreement among language maintainers was that we didn't want to collectively share the burden of maintaining it (when making changes that affect ocamlnat somehow), it should be the people who want to keep it alive that do that work (by fixing it after-the-fact). That may explain why ocamlnat-related tests are disabled: by that policy we want to be able to merge commits that do break ocamlnat tests.

If we want to stick to that policy, then the ocamlnat tests could run but should not be blocking -- I guess we would need a category of "ignorable" tests. Seeing the failures might motivate people to fix them earlier than today, and one could think of a `--really-test-ocamlnat` configure-level option for people to turn them into breakage if they care.

If we want to revisit the policy and ask that every changes to the compiler distribution preserves the functioning of ocamlnat, we need to have a discussion about it.

**Comment by gasche** [comment]
(cc @trefis)

**Comment by Armael** [comment]
I think having a category of tests that are allowed to fail is useful generally speaking. 
I was thinking about that in the context of @trefis location-related tests (#8987): it would be nice to have tests that *fail* when the parsetree does not satisfy the location-related invarints---but allow these tests to fail for the moment because it is too much work to fix them at once.

**Comment by xavierleroy** [comment]
@gasche spoke wisely. Also, if ocamlnat is tested, will it be tested on all the target architectures?  It's unclear to me whether ocamlnat works on all the target architectures, or whether we want it to.

**Comment by dra27** [comment]
Another option might be to enable them on one of the Inria CI jobs with a separate notification? It’s probably useful to know when it failed.

Given that it does have users, it might also be worth enabling the tests on the maintenance branches - ie if someone ensures ocamlnat worked when the version was branched, we all commit to keeping it working during beta?

**Comment by shindere** [comment]
My initial motivation for testing ocamlnat was the idea that, if a
behaviour varies between ocaml and ocamlnat, it may be the sign of
something that should / oculd have been factorised and was not.

**Comment by Armael** [comment]
(wrt allowing tests to fail, and following an offline suggestion by @shindere)

Instead of allowing tests to fail, a better option from a user perspective point of view might be to allow tests to emit *warnings*. For instance, the "location-invariants" test would always succeed, but output a list of warnings, that we then could decide to address. 
I think that makes much more sense that what I was initially suggesting.

**Comment by shindere** [comment]
David Allsopp (2019/09/27 07:18 -0700):
> Another option might be to enable them on one of the Inria CI jobs
> with a separate notification?

The extra-checks job might be a good place, but then there is the risk
that failures will be detected only after PRs have been merged, which,
as we all know, is a burden.

**Comment by dra27** [comment]
@shindere wrote:
> >  Another option might be to enable them on one of the Inria CI jobs
> >  with a separate notification?

> The extra-checks job might be a good place, but then there is the risk
> that failures will be detected only after PRs have been merged, which,
> as we all know, is a burden.

In this case it was intentional - the idea was that it would be a *separate* notification, so only going to anyone who is concerned about maintaining `ocamlnat`.

**Comment by Armael** [comment]
Closing, since following the discussion on #8911 it seems at the moment we do not want to spend any resources on maintaining ocamlnat (I was not aware of that).
